### PR TITLE
Finish TODOs left from axis domain computation

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

I want this selector to eventually replace the axisMap from generateCategoricalChart, I think we're close!

## Related Issue

https://github.com/recharts/recharts/issues/4583

## Motivation and Context

Direct DOM access

## How Has This Been Tested?

npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

File: src/state/dataSelectors.ts [hunk 1]
```diff
@@ -16,9 +18,25 @@ export const selectChartDataWithIndexes = (state: RechartsRootState): ChartDataS
...
+  axisId: AxisId,
+) => ReadonlyArray<ChartData> = createSelector(
+  (state: RechartsRootState) => state.graphicalItems.cartesianItems,
+  (_state, axisType: AxisType) => axisType,
+  (_state, _axisType, axisId: AxisId) => axisId,
+  (cartesianItems: ReadonlyArray<CartesianGraphicalItemSettings>, axisType: AxisType, axisId: AxisId) =>
+    cartesianItems
+      .filter(item => {
+        if (axisType === 'xAxis') {
+          // This is sensitive to the data type, as 0 !== '0'. I wonder if we should be more flexible. How does 2.x branch behave? TODO write test for that
```
Comment (ckifer) [inline src/state/dataSelectors.ts:33]: I think dataType matters in 2.x as well

File: src/state/dataSelectors.ts [hunk 2]
```diff
@@ -16,9 +18,25 @@ export const selectChartDataWithIndexes = (state: RechartsRootState): ChartDataS
...
+  axisType: AxisType,
+  axisId: AxisId,
+) => ReadonlyArray<ChartData> = createSelector(
+  (state: RechartsRootState) => state.graphicalItems.cartesianItems,
+  (_state, axisType: AxisType) => axisType,
+  (_state, _axisType, axisId: AxisId) => axisId,
+  (cartesianItems: ReadonlyArray<CartesianGraphicalItemSettings>, axisType: AxisType, axisId: AxisId) =>
+    cartesianItems
+      .filter(item => {
+        if (axisType === 'xAxis') {
```
Comment (ckifer) [inline src/state/dataSelectors.ts:32]: what about y axes? or if the chart is vertical?

File: src/state/graphicalItemsSlice.ts [hunk 3]
```diff
@@ -10,18 +22,18 @@ export type GraphicalItemsState = {
    */
   countOfBars: number;
   /**
-   * An array of arrays; the top array is for graphical items,
-   * one graphical item, one record in the array.
-   * The second level is because every data={x} on every item is expected to be an array itself.
+   * This is an array of all graphical items and their settings.
+   * Graphical item is a visual representation of data on the chart.
+   * Some examples are: Pie, Line, Bar.
```
Comment (ckifer) [inline src/state/graphicalItemsSlice.ts:27]: isn't Pie technically polar?

File: src/state/graphicalItemsSlice.ts [hunk 4]
```diff
@@ -34,15 +46,19 @@ const graphicalItemsSlice = createSlice({
...
     },
-    addGraphicalItemData(state, action: PayloadAction<ChartData>) {
-      state.graphicalItemData.push(action.payload);
+    addCartesianGraphicalItem(state, action: PayloadAction<CartesianGraphicalItemSettings>) {
+      state.cartesianItems.push(castDraft(action.payload));
     },
-    removeGraphicalItemData(state, action: PayloadAction<ChartData>) {
-      state.graphicalItemData = state.graphicalItemData.filter(item => item !== action.payload);
+    removeCartesianGraphicalItem(state, action: PayloadAction<CartesianGraphicalItemSettings>) {
+      const index = current(state).cartesianItems.indexOf(action.payload);
```
Comment (ckifer) [inline src/state/graphicalItemsSlice.ts:53]: why do we need `current`?

File: test/util/isDomainSpecifiedByUser.spec.ts [hunk 5]
```diff
@@ -100,36 +100,42 @@ describe('parsing axis domain provided by user', () => {
     { domain: [0, 100], expected: [0, 100] },
     { domain: [100, 100], expected: [100, 100] },
     { domain: [() => 1, 9], expected: [1, 9] },
+    /*
+     * I am surprised, but in Recharts it is valid to have domain where first element is larger than the second!
+     * This also makes the axis reversed.
+     */
+    { domain: [2, 1], expected: [2, 1] },
```
Comment (ckifer) [inline test/util/isDomainSpecifiedByUser.spec.ts:107]: lol 😅

**Review by ckifer — APPROVED** [review]

File: src/polar/Pie.tsx [hunk 6]
```diff
@@ -727,7 +726,8 @@ export class Pie extends PureComponent<Props, State> {
        */
       return (
         <>
-          <SetGraphicalItemData dataDefinedOnItem={this.props.data} />
+          {/* TODO introduce the polar alternative */}
+          {/* <SetCartesianGraphicalItem data={this.props.data} /> */}
```
[the remark this answers is not in the document]
Comment (PavelVanecek) [reply src/polar/Pie.tsx:730]: Correct. To-do

File: src/state/dataSelectors.ts (hunk 2 above)
Comment (PavelVanecek) [reply src/state/dataSelectors.ts:32]: To-do

File: src/state/graphicalItemsSlice.ts (hunk 3 above)
Comment (PavelVanecek) [reply src/state/graphicalItemsSlice.ts:27]: Ah yes I need to update the comment.

File: src/state/graphicalItemsSlice.ts (hunk 4 above)
Comment (PavelVanecek) [reply src/state/graphicalItemsSlice.ts:53]: Immer wraps objects in a Proxy. I want to remove state elements based on comparing the reference so I need a reference to the original state, not the proxy. So, current.

Redux recommendation is to use some sort of ID. But Recharts doesn't use IDs (except axes) so we need current.

Comment (ckifer) [reply src/state/graphicalItemsSlice.ts:53]: Got it, thank you!

**Pull request merged**